Fixed bug in Fast Adjacency Search (FAS)

MarcelRobeer opened the pull request:
Had an issue with running out memory in my own code and adjusting the `depth` parameter of `FCI` did not change anything. Traced it back to the _Fast Adjacency Search_ (FAS) step and found that depth is set to -1 regardless of what the user provided as input.

**Fix**: depth was always set to -1 regardless of provided parameter to function `causallearn.utils.Fas.fas()`, is now the user-provided depth.

wean2016 commented:
Dear @MarcelRobeer @chenweiDelight @kunwuz ,

Thanks for MarcelRobeer's critical PR and both Wei and Yujia's review.

I recently reviewed this PR. I agree with MarcelRobeer's opinion that there is some wrong with the parameter 'depth' of function 'fas'. However, I think MarcelRobeer's updated version is wrong. Specifically, the parameter 'depth' of 'fas' is NOT used in this version. In addition, the variable 'depth' in the content of 'fas' is an index for a 'while roop'.  In another word, the parameter 'depth' is not the same one in the current content of 'fas'. As a result, if we simply align these two variables, we will get the wrong algorithm result.

I think maybe one of the correct updated version can be described as follow. If I am wrong, please tell me. First, change the parameter 'depth' to 'max_search_depth'. Then, replace 'while cg.max_degree() - 1 > depth:' with 'while cg.max_degree() - 1 > depth and depth < max_search_depth:' in line 67. 

At last, I believe that @MarcelRobeer, who first finds this BUG, should own this committed contribution. So, can you further modify your submission version for subsequent acceptance?

Yuequn Liu,
2023.3.15

kunwuz commented:
Thanks so much for all the contributions, @MarcelRobeer, @wean2016 and @chenweiDelight ! @wean2016 could you please kindly review the recent commits, when you have time？

wean2016 commented:
Hi, everyone. Thanks very much for @MarcelRobeer. The new version is high quality! Due to lots of updated points, maybe some unit tests can help us verify the correctness. We appreciate it if @MarcelRobeer can further propose the running result of 'TestFCI.py' for the submitted version.

MarcelRobeer commented:
There were some issues with signing prior commit, so I commit f4c3771 contains all changes in one single commit:
1. `causallearn.utils.Fas` was renamed to `causallearn.utils.FAS`.
2. `fas()`: Simpler behavior for `tqdm` (no manual updating required at multiple places).
3. `fas()`: Replaced superfluous construction of intermediate variables (e.g. booleans, lists that are only later used as sets, ...).
4. `fas()`: Inputs are checked for their type correctness and throw a TypeError if the wrong type of input was provided.
5. `fas()`: Now returns the test scores of the independence tests with `test_results` (dictionary containing tests for x _||_ y | S and their p-value). These are required for testing the method.
6. Added unit tests to `tests.TestFAS` (largely based on `tests.TestFCI`) that test `fas()` for (1) input type sanity, (2) background knowledge forbidden edges not appearing in the resulting graph, and (3) if the depth parameter is applied correctly

wean2016 commented:
Thanks @MarcelRobeer for the awesome work! It looks perfect. I think this PR could be accepted for the main branch. @kunwuz @chenweiDelight

kunwuz reviewed (commented):
Thanks so much! Just one more quick commit and I think it will be good to go.

MarcelRobeer commented on dist/causal_learn-0.1.3.3-py3.10.egg:
Will fix this! Also made #104 so that in the future these will be prevented from being pushed.
